Add Gateway Docs (#280)
ethersphere/bee-docs: docs/installation/gateways.md

---
title: Swarm Gateways
id: gateways
---

Swarm is a p2p network application, like a Blockchain, and, in order to access it, you must run a node which will connect you directly to other user's computers across the globe.

At present, it is not yet possible to run a node directly in the browser and other web2 scenarios.

To overcome this, Swarm provides HTTP **Gateways** which can be accessed directly with a traditional web browser. These, however, provide only some of the functionality that Swarm offers.

## Run Your Own Gateway

There are three main types of gateway:

### Public

A public gateway is provided for the general public to access any data from Swarm. The Swarm Foundation and Fair Data Society provide [some of these](/docs/access-the-swarm/gateway), for you to try.

> We welcome those who would like to contribute to the decentralisation and health of the network to run their own public gateways. However, we recommend that only people and institiutions that are aware of the potential risks and legal liabilities of allowing anyone to access any content through a Swarm gateway they operate should attempt this.

Reviewer: Typo
```suggestion
We welcome those who would like to contribute to the decentralisation and health of the network to run their own public gateways. However, we recommend that only people and institutions that are aware of the potential risks and legal liabilities of allowing anyone to access any content through a Swarm gateway they operate should attempt this.
```

---
title: Swarm Gateways
id: gateway
---

> For now, not all of your devices, friends and users of your app will be able to run a Bee node which is able to connect directly to the network using *libp2p*. 

Reviewer 2: ```suggestion
For now, not all of your devices, friends and users of your app will be able to run a Bee node which is able to connect directly to the Swarm network. 
```
IMHO the `libp2p` is a technicality most users will not know about

> To access the Swarm network in these situations, it may be useful for you to provide a Swarm Gateway, which enables devices with only access to traditional web2 TCP/HTTP connections access to some elements of the swarm.

Reviewer 2: ```suggestion
To access the Swarm network in these situations, it may be useful for you to provide a gateway, which enables devices with access to only traditional web2 TCP/HTTP connections to access some elements of the swarm.
```

> 

Reviewer 3: to access

## Public Gateways

The Swarm Foundation provide access to various public gateways.

### Swarm Foundation Gateway

> Use the [Swarm Gateway](https://gateway.ethswarm.org/) to try out Swarm, you are able to upload up to 10mb at a time for free. Upload files, directories, or even folders containing websites with and `index.html` file to host them on Swarm.

Reviewer 2: ```suggestion
Use the [Swarm Gateway](https://gateway.ethswarm.org/) to try out Swarm, you are able to upload up to 10 MB at a time for free. Upload files, directories, or even folders containing websites with an `index.html` file to host them on Swarm.
```

You will receive an access link you can then share with others who will be able to retrieve the uploaded file, or browse your website directly from any gateway.

`https://gateway.ethswarm.org/`

### Swarm Foundation Gateway API

> If you would like to try out *swarm-cli* or *bee-js* you can start by specfiying the public gateway for operations that do not require access to the *Debug API*.

Reviewer 2: ```suggestion
You can also use the public gateways to access Swarm through *swarm-cli* or *bee-js*. This may be sufficient for some use cases, but many will require access to *Debug API*, which is not exposed on any gateway.
```

> 

Author: i'll provide links

`https://api.gateway.ethswarm.org/`

### BZZ.link

Use [BZZ.link](https://bzz.link/) to share links using CID's or ENS hosted websites on [Swarm](https://swarm.bzz.link/) in a secure browser context.

> 

Reviewer 3: and what is bzz.link?

`https://bzz.link`

> 

Reviewer 3: so what is the gateway api?

## Limitations

### Rate and Filesize Limits

To ensure everyone is able to try out the Swarm network, rate and filesize limitations are imposed on these public gateways. If you require that yourself, or the users of your Dapp require access which exceeds these limits, you must *run your own gateway*.

### Postage Stamps

When uploading chunks to Swarm, you must provide each with an attached signed postage stamp. Using our public gateway or API, chunks are automatically stamped with random batches that last for about 2 weeks. There is no way to extend or enlarge these batches, or to pick which one you will use.

> If you want to be able to control the amount of time your content stays in Swarm, you must have knowledge of your *Batch ID*. In order to choose which chunks are stamped with which batches, you must have access to the relevant key material. Since browser based chunk stamping is not yet implemented in javascript, Gateway Proxy must handle stamping chunks.

Reviewer 3: is Batch ID related to postage stamps? How should i know?
stamped with batches?
access to key material?
and what if i understood all this, I have the key material and know my batch ID, what do i do?

this is completely cryptic to someone not in the know.

> 

Reviewer 3: the Gateway Proxy must handle the stamping of chunks ?? no?

Reviewer 2: It of course does not have to, in the end it is the Bee node behind it that does the stamping. `gateway-proxy` just enables stamp management so that the end user does not have to do it. I am happy to help rephrase this. Please see the modes of operation for `gateway-proxy` https://github.com/ethersphere/gateway-proxy#usage

## Run Your Own Gateway

> Swarm does not yet have a browser based lightnode component, although this feasible and is planned for the future.

Reviewer 3: what is a lightnode component?

this IS feasible

Reviewer 2: ```suggestion
Swarm does not yet have a client that can be run in browser, although this is feasible and is planned for the future.
```

> In the meantime, in order to access Swarm directly from all your devices, it is possible for you to set use a HTTP gateway. This is similiar to using an RPC provider like *Infura* or *GetBlock* instead of running a blockchain node.

Reviewer 3: set use -> set or use

Reviewer 2: ```suggestion
In the meantime, in order to access Swarm directly from all your devices, it is possible for you to use a gateway. This is similar to using an RPC provider like *Infura* or *GetBlock* instead of running a blockchain node.
```

There are three main ways to run a gateway:

### Personal

> A personal gateway is to provide Swarm for you and your trusted colleagues, friends and family. This should be protected so that only authenticated users can access it. You can do this by enabling token based authentication, or in the safety of a private network.

Reviewer 3: provide an entire swarm network for myself and friends?
this can easily be misunderstood

### Public

> A public gateway is provided for the general public to access any data from Swarm. Only people and institiutions that are aware of the potential risks of allowing anyone to access a Swarm gateway they operate should attempt this.

Reviewer 3: so am I one? what are the potential risks?

### Dapp Provider Gateway

> Deploy your Dapp to a Web2 domain and provide an associated swarm gateway. You can use CORS to block* access from other websites using your gateway.

Reviewer 3: ??? deploy my dapp to a web2 domain? 
provide an associated swarm gateway?
how? what do i do?

> To deal with data persistence, there are three approaches which can be chosen based on how long you need your user's content to stay available in the network:

Reviewer 3: user's content -> users' content

> - enable automatic stamping for short lived files

Reviewer 3: how?

- enable automatic stamping using the same large batch, by setting a default `POSTAGE_STAMP`

> - pass any postage batch id as a header. the postage batch must belong to your attached Bee node.

Reviewer 3: how ? what does it mean a batch belongs to a node?

* this is only blocked at the browser level, you may want to otherwise authenticate your users sessions or limit access to a subset of the swarm network.

> ** we are currently investigating auto top up for the default stamp, for now please manage this manually using *swarm-cli* or *swarm desktop*.

Reviewer 3: auto top-up? 
per-user stamp ids?

Reviewer 2: I would remove these two sentences. We can create github issues instead and see if there is an interest in implementing them.

*** we are currently investigating ideas on automatically generating per user stamp id's using the Gateway Proxy.

## Overview 

In order to run a gateway, you must:

> - configure DNS records for your gateway's domain.

Reviewer 3: are these gonna be explained ?

- "BEE_VERBOSITY=debug"

Consider setting your Bee node's address if it has not been automatically connected so that other nodes in the network can discover it.

- "BEE_NAT_ADDR=123.123.123.123:1634"

When you start up your node, check the logs. You will see something like:

```
cannot continue until there is at least 0.000250 xDAI (for Gas) available on b33b33b33b33b33b33b33b33b33b33b33b33b333
```

Fund the address with xDAI. Once your have done so, your node should begin to sync postage batch snapshot and once it has done so, connect to the network. 

## Testing your installation

If you have enabled ENS support:

1. https://swarm.my-domain.org should resolve to the Swarm website.
2. https://my-domain.org/bzz/swarm.eth should resolve to the Swarm website.

### Modes of Operation

Follow the relevant instructions to get your gateway up and running to suit your needs.

###  Personal

Supply the following configuration parameters to Gateway Proxy in order to create an authenticated gateway that automatically stamps chunks with a default stamp.

Limit access to your gateway by providing an authorisation secret known only to you and trusted individuals.

```

> AUTH_SECRET=b33b33b33b33b33b33b33b33b33

Reviewer 2: TBH I would rather recommend the `nginx` authentication than the build in one. We wanted to remove it because it is quite limited (there is only one secret for instance that you need to use for all applications).

Author: ok, let me see what i can do

```

### Public

The following configuration parameters will set up the gateway to autobuy stamps of a certain duration. Stamps will be automatically rotated when they are getting full or running out of TTL. See *Gateway Proxy* for more configuration parameters.

```
POSTAGE_DEPTH=19
POSTAGE_AMOUNT=1000000
```

### Dapp Provider Gateway

Use `swarm-cli`, to create a postage stamp then provide this to Gateway Proxy as the default. This will then be used to stamp all chunks created using the gateway api.

```
POSTAGE_STAMP=b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33b33
```

If deploying a browser based js app, you may want to serve up your applications's code at your domain root address. You can use swarm to do this by proxying the root route to your application's swarm address or ENS ref.

### Docker Compose Quick Start

Use the following script to get a gateway running using only one command.

First, configure the script below to customise it for your needs and copy it to your server, having installed the *docker* software.

1. Ensure you have configured DNS records as indicated above.
2. Create and account at Cloudflare and create an access token, configure:

>   - `CLOUDFLARE_EMAIL=your@email.com`

Reviewer 3: where do i configure these?

  - `CLOUDFLARE_DNS_API_TOKEN=b33b33b33b33b33b33b33b33b33b33b33`
3. Add your email for letsencrypt (certs are reprovisioned automatically).
  - `certificatesresolvers.gatewayresolver.acme.email=your@email.com`
3. Configure your domain name.
  - `entrypoints.websecure.http.tls.domains[0].main=my-gateway.org`
  - `entrypoints.websecure.http.tls.domains[0].sans=*.my-gateway.org`
  - `traefik.http.routers.gateway.rule=HostRegexp('my-gateway.org,'{subdomain:[a-z]+}.my-gateway.org')`
  - `HOSTNAME=swerm.org"`
4. Add your RPC endpoint addresses with API keys.
  - `BEE_RESOLVER_OPTIONS=https://mainnet.infura.io/v3/0b858795af534ad6a247181cc535e775`
  - `BEE_SWAP_ENDPOINT=https://gno.getblock.io/mainnet/?api_key=e2fd8ad8-cd39-4d31-8905-a673c6e8ec0b`
5. Optionally, add CORS protections.
  - `BEE_CORS_ALLOWED_ORIGINS[https://my-gateway.org]`

> 6. Optionally, add an authentication token.

Reviewer 3: these settings are all where?

                  - "traefik.http.middlewares.gatewaycorsheader.headers.accessControlAllowOriginListRegex=(.*?)"
                  - "traefik.http.middlewares.gatewaycorsheader.headers.accesscontrolmaxage=100"
                  - "traefik.http.middlewares.gatewaycorsheader.headers.addvaryheader=true"
                  - "traefik.http.routers.gatewayrouter.middlewares=gatewaycorsheader@docker"

  bee:
    image: ethersphere/bee:stable
    container_name: bee
    restart: unless-stopped
    volumes:
      - bee:/home/bee/.bee
    command: start
    environment:
            - "BEE_API_ADDR=:1633"
            - "BEE_P2P_ADDR=:1634"
            - "BEE_CORS_ALLOWED_ORIGINS=['*']"
            - "BEE_DEBUG_API_ADDR=:1635"
            - "BEE_DEBUG_API_ENABLE=true"
            - "BEE_NAT_ADDR"
            - "BEE_RESOLVER_OPTIONS=https://mainnet.infura.io/v3/b33b33b33b33b33"
            - "BEE_SWAP_ENDPOINT=https://gno.getblock.io/mainnet/?api_key=b333b33b3b3b3b3333"
            - "BEE_SWAP_INITIAL_DEPOSIT=0"
            - "BEE_PASSWORD=my-password"
            # - "BEE_VERBOSITY=debug"
    ports:
       - "1633:1633"
       - "1634:1634"
       - "1635:1635"

volumes:
  bee:

> ```

Reviewer 3: and now what? what if i run this? whats next?

### Kubernetes

> The K8s scripts we use in-house are not yet public, but our friendly devops team are more than happy to help you in any way they can! Get in touch via our *Discord* server.

Reviewer 3: what is k8s and why would i need it?
